Fix 'ansible-vault edit /some/symlink'

alikins opened the pull request:
Since vault edit attempts to unlink
edited files before creating a new file
with the same name and writing to it, if
the file was a symlink, the symlink would
be replaced with a regular file.

VaultEditor file ops now check if files
it is changing are symlinks and instead
works directly on the target, so that
os.rename() and shutils do the right thing.

Add unit tests cases for this case and
assorted VaultEditor test cases.

Fixes #20264

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->

 - Bugfix Pull Request
 - Test Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
lib/ansible/parsing/vault

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.3.0 (vault_edit_link_20264 c502a6e50b) last updated 2017/01/18 14:48:44 (GMT -400)
  config file = /home/adrian/ansible/ansible.cfg
  configured module search path = [u'/home/adrian/ansible/my-modules', u'/home/adrian/ansible/lib/modules/']

```

bcoca commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -526,32 +526,59 @@ def write_data(self, data, filename, shred=True):
-            with open(filename, "wb") as fh:
-                fh.write(b_file_data)
+            return
+
```
~~don't you still need the else? won't this fail when using stdin/stdout?~~ nvmd, missed the 'return'

bcoca commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -526,32 +526,59 @@ def write_data(self, data, filename, shred=True):
+            else:
+                os.remove(target_filename)
+
+        return target_filename, stat_info
```
~~why stat and return the data when you are dropping it on caller?~~ 2nd caller does use

bcoca commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -526,32 +526,59 @@ def write_data(self, data, filename, shred=True):
-                fh.write(b_file_data)
+            return
+
+        target_filename, _ = self._unlink_file(filename, shred=shred)
```
not sure we need this here as we should always be writing to a tmp file

alikins replied:
encrypt_file() and decrypt_file() call write_data(data, filename) with filename being a non-tmp file.

alikins replied:
afaict, the 'remove the file, then re-open and write to it' approach was added so that 'shred' could be called on the old file before writing the encrypted version. _unlink_file tries to preserve that,

bcoca commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -526,32 +526,59 @@ def write_data(self, data, filename, shred=True):
-            # old file 'dest' was encrypted, no need to _shred_file
-            os.remove(dest)
-        shutil.move(src, dest)
+        target_dest, prev = self._unlink_file(dest)
```
this creates race condition. there is a time between remove and move that the file is unavailable. I see original code did same, but we should just allow move to work as it will be an atomic operation

alikins replied:
It's definitely a race condition. The 'shred' hook kind of makes real atomic ops difficult. 

Not entirely convinced the 'shred' hooks are worth the complexity, but giving it the benefit of the doubt. The existing code seems to go out of it's way to always have a finally path to shred old files
and seems to value that over atomic file ops. If thats no longer the case, the code can be simplified and made more atomic[1]

For the non-shred cases, we could skip the os.remove() and let rename take care of it (via shutils.move[1]). 

[1] at least for the scenarios where os.rename() will work. Given the tempfile.mkstemp() use, I suspect os.rename() wont work for most cases (assuming /tmp of different fs...).

alikins commented:
Just pushed a variation that removes the attempts to shred from before the file writes, until after in a finally clause.

bcoca commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -500,58 +503,42 @@ def rekey_file(self, filename, new_password):
+
+        try:
+            with open(filename, "rb") as fh:
+                data = fh.read()
```
why do we need to separate the error handling?

alikins replied:
my pref is to have the special case (filename='-') checked early and to return early and have the normal case be in the first indent level in the method. Also minimizes the amount of code that the raised AnsibleErrors could be referring to.

bcoca commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -380,39 +379,42 @@ def _shred_file(self, tmp_path):
-            # whatever happens, destroy the decrypted file
+            b_tmpdata = edit_tmp_file(tmp_path)
+        except Exception as e:
+            print(e)
```
we normally use display instead of print

alikins replied:
Think I was intending to remove that except stanza entirely and forgot to remove that debugging stanza

bcoca commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -380,39 +379,42 @@ def _shred_file(self, tmp_path):
 
-        if existing_data:
-            self.write_data(existing_data, tmp_path, shred=False)
+        def edit_tmp_file(tmp_file_path):
```
is there any reason we need to encapsulate the code here?

alikins replied:
just to have it in one try/except/finally without the try being being 15 lines

bcoca commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -380,39 +379,42 @@ def _shred_file(self, tmp_path):
             # we could not successfully execute unix shred; therefore, do custom shred.
             self._shred_file_custom(tmp_path)
 
-        os.remove(tmp_path)
```
i think we require this

bcoca commented on lib/ansible/parsing/vault/__init__.py:
```diff
@@ -500,58 +503,42 @@ def rekey_file(self, filename, new_password):
+            return
+
+        with open(filename, "wb") as fh:
+            fh.write(b_file_data)
```
This introduces a different race condition, file can now be looked at in 'intermediate' state.

alikins replied:
Was trying to avoid the gap between the tmp file rename and resetting perms/attributes in shuffle_files. This does allow a read of a partially written file, but it also keeps file perms in place.

I like the idea of getting perms sorted on the tmpfile before the rename  (as mentioned at https://github.com/ansible/ansible/issues/18537). Though still a small perm race there (dest file perms could change in the gap between the setting/copying file perms from dest file to the temp file and when the rename happens) .

sivel commented:
I may have a misunderstanding here, but wouldn't it be sufficient to just do `os.path.realpath` on the provided filename, and do all operations on this new path?

That leaves us with only needing to make a single line change, to accommodate this.  A simple test of making the following change seemed to do as expected:

```
diff --git a/lib/ansible/parsing/vault/__init__.py b/lib/ansible/parsing/vault/__init__.py
index 3c7a89a..72425b6 100644
--- a/lib/ansible/parsing/vault/__init__.py
+++ b/lib/ansible/parsing/vault/__init__.py
@@ -448,7 +448,7 @@ class VaultEditor:
         self._edit_file_helper(filename)

     def edit_file(self, filename):
-
+        filename = os.path.realpath(filename)
         check_prereqs()

         ciphertext = self.read_data(filename)
```

sirkubax commented:
please close https://github.com/ansible/ansible/pull/20219 (as it's limited functionality compared to this PR) when you merge this PR

alikins committed:
Fix 'ansible-vault edit /some/symlink'

Since vault edit attempts to unlink
edited files before creating a new file
with the same name and writing to it, if
the file was a symlink, the symlink would
be replaced with a regular file.

VaultEditor file ops now check if files
it is changing are symlinks and instead
works directly on the target, so that
os.rename() and shutils do the right thing.

Add unit tests cases for this case and
assorted VaultEditor test cases.

Fixes #20264

alikins commented:
updated to just use realpath to follow links for encryt_file/decrypt_file/rekey_file/edit_file

The other preexisting bugs can be address in a different pr.
